src/WLR/Types/Keyboard.hsc

```diff
@@ -0,0 +1,169 @@
+{-# LANGUAGE EmptyDataDeriving, PatternSynonyms #-}
+
+module WLR.Types.Keyboard where
+
+#define WLR_USE_UNSTABLE
+#include <wlr/types/wlr_keyboard.h>
+
+import Foreign (Word32, Storable(..))
+import Foreign.C.Types (CSize(..), CInt(..), CBool(..))
+import Foreign.C.String (CString)
+import Foreign.Ptr (Ptr)
+
+import WLR.Types.InputDevice (WLR_input_device)
+import WLR.Types.KeyboardGroup (WLR_keyboard_group)
+
+import WL.ServerCore (WL_signal)
+import WL.Keyboard (WL_keyboard_key_state)
+
+pattern WLR_LED_COUNT :: (Eq a, Num a) => a
+pattern WLR_LED_COUNT = #const WLR_LED_COUNT
+
+type WLR_keyboard_led = CInt
+-- enum wlr_keyboard_led
+--WLR_LED_NUM_LOCK = 1 << 0,
+pattern WLR_LED_NUM_LOCK :: (Eq a, Num a) => a
+pattern WLR_LED_NUM_LOCK = 1
+--WLR_LED_CAPS_LOCK = 1 << 1,
+pattern WLR_LED_CAPS_LOCK :: (Eq a, Num a) => a
+pattern WLR_LED_CAPS_LOCK = 2
+--WLR_LED_SCROLL_LOCK = 1 << 2,
+pattern WLR_LED_SCROLL_LOCK :: (Eq a, Num a) => a
+pattern WLR_LED_SCROLL_LOCK = 4
+
+pattern WLR_MODIFIER_COUNT :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_COUNT = 8
+
+type Wlr_keyboard_modifier = CInt
+-- enum wlr_keyboard_modifier {
+-- WLR_MODIFIER_SHIFT = 1 << 0,
+pattern WLR_MODIFIER_SHIFT :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_SHIFT = 1
+-- WLR_MODIFIER_CAPS = 1 << 1,
+pattern WLR_MODIFIER_CAPS :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_CAPS = 2
+-- WLR_MODIFIER_CTRL = 1 << 2,
+pattern WLR_MODIFIER_CTRL :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_CTRL = 4
+-- WLR_MODIFIER_ALT = 1 << 3,
+pattern WLR_MODIFIER_ALT :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_ALT = 8
+-- WLR_MODIFIER_MOD2 = 1 << 4,
+pattern WLR_MODIFIER_MOD2 :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_MOD2 = 16
+-- WLR_MODIFIER_MOD3 = 1 << 5,
+pattern WLR_MODIFIER_MOD3 :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_MOD3 = 32
+-- WLR_MODIFIER_LOGO = 1 << 6,
+pattern WLR_MODIFIER_LOGO :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_LOGO = 64
+-- WLR_MODIFIER_MOD5 = 1 << 7,
+pattern WLR_MODIFIER_MOD5 :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_MOD5 = 128
+
+pattern WLR_KEYBOARD_KEYS_CAP :: (Eq a, Num a) => a
+pattern WLR_KEYBOARD_KEYS_CAP = 32
+
+data {-# CTYPE "wlr/types/wlr_keyboard.h" "struct wlr_keyboard impl" #-} WLR_keyboard_impl
+    deriving (Show)
+
+-- xkbd_mod_mask_t is a type alias for uint32_t
+
+{{ struct
+    wlr/types/wlr_keyboard.h,
+    wlr_keyboard_modifiers,
+    depressed, Word32,
+    latched, Word32,
+    locked, Word32,
+    group, Word32
+}}
+
+-- cannot import these types from libxcommon because of their fields which have
+-- internal types that aren't exported
```

Comment:
What do you mean by this? Regardless, looks like there's plenty of nested structs in each of those, so this is work that's best put off for another PR.

One request though; could you capitalize these as `XKB_keymap` and `XKB_state`? I think it's best to keep the naming consistent with what we've got already.

Reply:
Yes I will caps the prefix.

About the imports though, I don't think we can use hsc features to create the bindings for things that are internal to xkbcommon. As far as I can tell they cannot be imported in any way, we'd have to write out something by hand that lines up with what they have perfectly.

I wonder if the fact that they're not exported means that we won't ever need to peek or poke them, and we just need to be able to pass them through to other C functions.

Reply:
> things that are internal to xkbcommon

This is what I'm confused about. Could you give a few examples?

Reply:
For example, I cannot find importable declarations of `struct xkb_keymap {`

There is a definition in `src/keymap.h` in the libxkbcommon repo. In that file they have this section:
```
 /* Don't use compat names in internal code. */
#define _XKBCOMMON_COMPAT_H
#include "xkbcommon/xkbcommon.h"
```
that implies that the file is 'internal code'

When I tried adding xkbcommon to cabal and including things from that file, it fails. I think they are hiding struct definitions that they don't want to be a part of the public API

Reply:
> that implies that the file is 'internal code'

There are some functions that are declared in `keymap.h` which are local, but I assume that those are just the functions defined in `keymap-priv.c`. It looks like all of the functions in `keymap.c` are marked with `XKB_EXPORT`, so those should be fine to write bindings to.

From the initial comment I was asking about:

> internal types that aren't exported

This is what I was particularly confused about. While some functions may have hidden visibility, I don't think that this should affect any of our mappings to structs/enums, since those were never symbols to be exported in the first place.

Regarding the opaque structures, that's probably a good thing. Less work for us to do :)